workbench-core/base/src/aws/helpers/dynamoDB/dynamoDBService.ts

```diff
@@ -140,6 +142,27 @@ export default class DynamoDBService {
     return getter;
   }
 
+  /**
+   * Queries the DynamoDB table.
+   *
+   * @param params - optional object of optional properties to generate a query request
+   * @returns ExecuteQueryResult object
+   *
+   * @example Use this to execute a query to the DynamoDb table.
+   * ```ts
+   * const result = dynamoDBService.executeQuery({sortKey: 'value', eq: {N: '5'}});
+   * ```
+   */
+  public async executeQuery(params?: QueryParams): Promise<ExecuteQueryResult> {
```

Comment:
What's the difference between using this method and `query.execute`? Their name sounds very similar so I'm worried developers might be confused between the two method. What are your thoughts on changing the return type of the query class to return ExecutionQueryResult?

Reply:
@nguyen102 we decided not to modify `query.execute` right now to prevent any breaking changes in the code. Any new code will use this new method and `query.execute` will be deprecated.

My overall plan is to remove all references to `DynamoDBService` in the code except where we're instantiating services. All the services that need to talk to dynamo will take the `DatabaseService` interface as a constructor argument, giving us a simplistic form of Dependency Injection. Any time someone needs to talk to the database, they can use the object passed in via the interface. We can add new methods as needed to `DatabaseService` and do a seamless refactor of any place `DynamoDBService` is still explicitly being used as we touch those parts of the code.

Reply:
Sorry, I forgot to answer what the difference is: 

Currently, `query.execute()` modifies the underlying objects in the `QueryCommandOutput`, specifically `QueryCommandOutput.Items` and `QueryCommandOutput.LastEvaluatedKey`, changing their values to `Record<string, any>` at run time, making the Typescript compiler basically useless because it can't calculate types accurately anymore. That swap also leads us to do a lot of manual type casting using `as unknown as Type`, which again effectively nullifies the typing system. The new method and gradual refactor mentioned above are meant to limit the blast radius of this issue and allow us to get back to having a reliable type checker.

Reply:
Ah thanks for explaining. I think it makes sense to create an `DatabaseService` interface. This will help with dependency injection as you mentioned. However, I think we need to add some major changes to the interface and clearly describe how we want the interface to be used if we're deciding to go with that route for now.

`DatabaseService` interface shouldn't be tied to DDB implementation details. For example, in the current interface the parameter value is [QueryParams](https://github.com/aws-solutions/solution-spark-on-aws/blob/develop/workbench-core/base/src/constants/queryParams.ts), which is using `AttributeValue`. `AttributeValue` is a DDB type definition and ties `DatabaseService` interface with DDB types.

The interface shouldn't be DDB specific. I'm thinking the interface should look something like this

```
interface DatabaseService {
    listDBResources<T>(param: ListDBResourceRequest, expectedResourceType: T) : ListDBResourceResponse<T>
}
```

```
interface ListDBResourceResponse <T> {
   data: T[], 
   paginationToken?: string
}
```

The syntax might need some massaging, but the high level concept is calling `listDBResources` should return a list of resources that are of the same type as `expectedResourceType`. `listDBResource` should check the resources that are return from the DB and confirm that they are the wrong type.

Without this check, we're still not checking the object type and we're still in the same state as before.
This line of code `const data = result.Items?.map((item) => item as unknown as Record<string, JSONType>) || [];` is doing the same thing as what we're already doing in the other services.